metrics/exporter.go

```diff
@@ -32,11 +33,107 @@ type flushable interface {
+	//    <domain>/<component>/<metric name from View>
+	// Prometheus uses the following format to construct full metric name:
+	//    <component>_<metric name from View>
+	// Domain is actually not used if metrics backend is Prometheus.
+	Domain string
+
+	// Component is the name of the component that emits the metrics. e.g.
+	// "activator", "queue_proxy". Should only contains alphabets and underscore.
+	// Must be present.
+	Component string
+
+	// PrometheusPort is the port to expose metrics if metrics backend is Prometheus.
+	// It should be between maxPrometheusPort and maxPrometheusPort. 0 value means
+	// using the default 9090 value. If is ignored if metrics backend is not
+	// Prometheus.
+	PrometheusPort int
+
+	// ConfigMap is the data from config map config-observability. Must be present.
+	// See https://github.com/knative/serving/blob/master/config/config-observability.yaml
+	// for details.
+	ConfigMap map[string]string
+}
+
+// UpdateExporterFromConfigMap returns a helper func that can be used to update the exporter
+// when a config map is updated.
+func UpdateExporterFromConfigMap(component string, logger *zap.SugaredLogger) func(configMap *corev1.ConfigMap) {
+	domain := Domain()
+	return func(configMap *corev1.ConfigMap) {
+		UpdateExporter(ExporterOptions{
+			Domain:    domain,
+			Component: component,
+			ConfigMap: configMap.Data,
+		}, logger)
+	}
+}
+
+// UpdateExporter updates the exporter based on the given ExporterOptions.
+func UpdateExporter(ops ExporterOptions, logger *zap.SugaredLogger) error {
+	metricsMux.Lock()
+	defer metricsMux.Unlock()
+
+	newConfig, err := createMetricsConfig(ops, logger)
+	if err != nil {
+		if curMetricsExporter == nil {
+			// Fail the process if there doesn't exist an exporter.
+			logger.Errorw("Failed to get a valid metrics config", zap.Error(err))
+		} else {
+			logger.Errorw("Failed to get a valid metrics config; Skip updating the metrics exporter", zap.Error(err))
+		}
+		return err
+	}
+
+	if isNewExporterRequiredHelper(newConfig, curMetricsConfig) {
+		logger.Info("Flushing the existing exporter before setting up the new exporter.")
+		flushExporterHelper(curMetricsExporter)
+		e, err := newMetricsExporterHelper(newConfig, curMetricsExporter, logger)
+		if err != nil {
+			logger.Errorf("Failed to update a new metrics exporter based on metric config %v. error: %v", newConfig, err)
+			return err
+		}
+		existingConfig := curMetricsConfig
+		setCurMetricsExporterUnlocked(e)
+		logger.Infof("Successfully updated the metrics exporter; old config: %v; new config %v", existingConfig, newConfig)
+	}
+
+	setCurMetricsConfigUnlocked(newConfig)
+	return nil
+}
+
+// isNewExporterRequired compares the non-nil newConfig against curMetricsConfig. When backend changes,
+// or stackdriver project ID changes for stackdriver backend, we need to update the metrics exporter.
+func isNewExporterRequired(newConfig *metricsConfig) bool {
+	cc := getCurMetricsConfig()
```

Comment:
Sorry I don't under standard this comment. What do you mean?

Reply:
isNewExporterRequired existed before this PR and is used by many of the metrics test. getCurMetricsConfig() here uses a lock, so to make sure all the tests passed w/o modification, I left the behavior of isNewExporterRequired untouched.

Reply:
I can see this as a temporal, but this has to be cleaned up.

Reply:
What would be your ideal form of cleanup? I could see moving the functions only used by tests to a separate file.

I could also remove the functions only used by tests completely and remove tests designed to test those functions.

Reply:
Basically having a function whose only purpose is to just call another function with a param seems like overhead. I'd like it to go away :)

Reply:
Ok, I've consolidated the two half-duplicated functions to just one unlocked function each. This is a change in existing behavior because they are no longer thread-safe. I've added a line in the function comment to point that out.

Tests are still fine because they are single-threaded.